library/pkcs7.c

```diff
@@ -0,0 +1,525 @@
+/* Copyright 2019 IBM Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+#if defined(MBEDTLS_PKCS7_USE_C)
+
+#include "mbedtls/x509.h"
+#include "mbedtls/asn1.h"
+#include "mbedtls/pkcs7.h"
+#include "mbedtls/x509_crt.h"
+#include "mbedtls/x509_crl.h"
+#include "mbedtls/oid.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#include <stdlib.h>
+#define mbedtls_free      free
+#define mbedtls_calloc    calloc
+#define mbedtls_printf    printf
+#define mbedtls_snprintf  snprintf
+#endif
+
+#if defined(MBEDTLS_HAVE_TIME)
+#include "mbedtls/platform_time.h"
+#endif
+#if defined(MBEDTLS_HAVE_TIME_DATE)
+#include "mbedtls/platform_util.h"
+#include <time.h>
+#endif
+
+/*
+ * Load all data from a file into a given buffer.
+ *
+ * The file is expected to contain DER encoded data.
+ * A terminating null byte is always appended.
+ */
+int mbedtls_pkcs7_load_file( const char *path, unsigned char **buf, size_t *n )
```

Comment:
This function isn't specific to PKCS7 and shouldn't be part of the public API.

Reply:
I am sorry I am not clear on this. I do see similar public API in pk.h
` "int mbedtls_pk_load_file( const char *path, unsigned char **buf, size_t *n );" `

And if I don't make it public API, how do I call it ? Did you mean there is some common public function for loading file ?

Thanks & Regards,
    - Nayna

Reply:
I have two questions on that function (on top of Hanno's):
1. Why does it append a 0 byte? If the data is expected is be DER, then I don't see any reason for this.
2. Why not use `mbedtls_pk_load_file()` instead? Unless I'm mistaken, PKCS7 depends on X509 which depends on PK, so this wouldn't introduce a new dependency. If use `mbedtls_pk_load_file()` is not desirable for some reason, this reason should be explain in a comment here.

In reponse to your question, when we created `mbedtls_pk_load_file()` we didn't have a way to make function usable by other library modules while not being part of the public API, but now we do: we declare them in a header in `library/xxx.h` instead of `include/mbedtls/xxx.h`. (See for example `library/ssl_tls13_keys.h`.) This if probably what we'd do with `mbedtls_pk_load_file()` if we were introducing it today. (Of course it was only natural for you to mimick what was done in PK, you couldn't know we'd changed our practice in the meantime.)